src/pybind/mgr/dashboard/services/auth/auth.py

```diff
@@ -25,6 +30,74 @@
 })
 
 
+class AuthType(str, Enum):
+    LOCAL = 'local'
+    SAML2 = 'saml2'
+    OAUTH2 = 'oauth2'
+
+
+class BaseAuth(abc.ABC):
+    LOGIN_URL: str
+    LOGOUT_URL: str
+    sso: bool
+
+    @staticmethod
+    def from_protocol(protocol: AuthType) -> Type["BaseAuth"]:
+        for subclass in BaseAuth.__subclasses__():
+            if subclass.__name__.lower() == protocol:
+                return subclass
+            for subsubclass in subclass.__subclasses__():
+                if subsubclass.__name__.lower() == protocol:
+                    return subsubclass
+        raise ValueError(f"Unknown auth backend: '{protocol}'")
+
+    @classmethod
+    def from_db(cls, db: Optional['SsoDB'] = None) -> Type["BaseAuth"]:
+        if db is None:
+            protocol = mgr.SSO_DB.protocol
+        else:
+            protocol = db.protocol
+        return cls.from_protocol(protocol)
+
+    class Config(TypedDict):  # pylint: disable=inherit-non-class
+        pass
+
+    @abc.abstractmethod
+    def to_dict(self) -> 'Config':
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def from_dict(cls, s_dict) -> 'BaseAuth':
+        pass
+
+    @classmethod
+    def get_auth_name(cls):
+        return cls.__name__.lower()
```

Comment:
Why is this defined in every subclass?
```py
>>> class A:
...   @classmethod
...   def name(cls):
...     return cls.__name__.lower()
... 
>>> A.name()
'a'
>>> class B(A): pass
... 
>>> B.name()
'b'
```

Reply:
OOP noob I guess...
First time I tried it I was getting an error ( don't remember which one), so I ended up with that.
Something must have been messed up in the inhertance before.

Anyways, I have updated it to remove the repetition